### ISSUE TITLE
Different behaviour when running test with async/await and without

### ISSUE THREAD

**Issue by macpak** [xunit/xunit#2111]
Hi,

My setup:

- Jetbrains Rider 2019.2.3
- Win 10
- .Net Core 3.1
- "maxParallelThreads": 1
- xunit 2.4.1
- xunit.runner.visualstudio 2.4.2

I've the following code:

```
using System;
using System.Threading.Tasks;
using Serilog;
using Xunit;
using Xunit.Abstractions;
using Xunit.Sdk;

[assembly: Xunit.TestFramework("XUnitAsyncIssue.MyClassName", "XUnitAsyncIssue")]

namespace XUnitAsyncIssue
{
    public class MyClassName : XunitTestFramework
    {
        public MyClassName(IMessageSink messageSink)
            :base(messageSink)
        {
            // Place initialization code here
            Log.Logger = new LoggerConfiguration()
                .WriteTo.Console()
                .WriteTo.File("log.txt")
                .CreateLogger();
        }

        public new void Dispose()
        {
            // Place tear down code here
            base.Dispose();
        }
    }
    
    public class Class1
    {
        [Fact]
        public async Task Test1()
        {
            Log.Logger.Information("Before test1");
            await Task.Delay(1000);
            Log.Logger.Information("End test1");
        }
    }
    
    public class Class2
    {
        [Fact]
        public async Task Test2()
        {
            Log.Logger.Information("Before test2");
            await Task.Delay(1000);
            Log.Logger.Information("End test2");
        }
    }
}
```
When I look at the output of the logger, I see logs that look like:
```
2020-06-23 20:11:18.041 +02:00 [Information] Before test2
2020-06-23 20:11:18.061 +02:00 [Information] Before test1
2020-06-23 20:11:19.068 +02:00 [Information] End test1
2020-06-23 20:11:19.068 +02:00 [Information] End test2
```
So basically, it seems that the code for the second test starts running before the first has finished.

The situation looks differently, when I remove async/await:
```
  public class Class1
    {
        [Fact]
        public void Test1()
        {
            Log.Logger.Information("Before test1");
            Thread.Sleep(1000);
            Log.Logger.Information("End test1");
        }
    }
    
    public class Class2
    {
        [Fact]
        public void Test2()
        {
            Log.Logger.Information("Before test2");
            Thread.Sleep(1000);
            Log.Logger.Information("End test2");
        }
    }
```
Now, I see the logs that look like:
```
2020-06-23 20:14:06.030 +02:00 [Information] Before test2
2020-06-23 20:14:07.060 +02:00 [Information] End test2
2020-06-23 20:14:07.075 +02:00 [Information] Before test1
2020-06-23 20:14:08.086 +02:00 [Information] End test1
```
So this time the logs seem to be ok - test2 starts/ends and then the other starts/ends. 
I'd expect to see the same logs for async/await scenario.

**Comment by bradwilson** [comment]
TL;DR: This is expected behavior. What you wanted to `"parallelizeTestCollections": false`, not `"maxParallelThreads": 1`.

------

What you've done is _not_ turning off parallelization; you've merely said that "at any given time, only one test may be running". When you yield the thread with your `await Task.Delay` call, you're returning the single thread to the pool to be used by another test. The system is allowed to start as many tests in parallel as possible to best use the available threads, and when a thread is no longer running because it's waiting (in your case, for `Task.Delay`, but more commonly tests will be waiting for async I/O), another test that's ready to run will start running. The purpose of this design is to maximize utilization of the CPU cores.

Your second example doesn't yield the thread (`Thread.Sleep` continues to consume the thread while sleeping). This is why it behaves like you expected.

To disable all parallelization in your test assembly, you should use `"parallelizeTestCollections": false`.

**Comment by macpak** [comment]
Thanks @bradwilson , makes sense.

The problem I've is that I'd like to have only n tests run at the given time. The reason for that is that we use Selenium with XUnit to run our tests - unfortunately, we've problems (with CI agents) if there are too many active sessions at the same time. So we used `"maxParallelThreads": 2` (we need parallelization as tests take much time to run) which worked ok until we've started using async/await inside the tests which resulated in more than 2 session being opened at the same time.
Is there a way to tell XUnit that I want to have max n tests run at the same time ?

**Comment by bradwilson** [comment]
Your only recourse is to stop using async in the unit tests.

**Comment by macpak** [comment]
I see, thanks. I assume we can close it.
